addons/sale_pdf_quote_builder/wizards/sale_pdf_quote_builder_dynamic_fields_wizard.py

```diff
@@ -0,0 +1,134 @@
+
+    dynamic_fields_wizard_id = fields.Many2one(
+        'sale.pdf.quote.builder.dynamic.fields.wizard', required=True, ondelete='cascade'
+    )
+    document_type = fields.Selection(
+        string="Document Type",
+        selection=[('header_footer', "Header/Footer"), ('product_document', "Product Document")],
+        default='header_footer',
+        required=True,
+    )
+    form_field = fields.Char(
+        string="Form Field", help="Form Field Name as seen in the PF form", required=True,
+    )
+    path = fields.Char(default='', help="Odoo path to get the expected value to fill the PDF.")
+
+    @api.constrains('form_field')
+    def _check_form_field_name_follows_pattern(self):
+        """ Ensure the paths only contains alphanumerics, hyphens and underscores.
+
+        :return: None
+        :raises: ValidationError if the names aren't alphanumerics, hyphens and underscores.
+        """
+        name_pattern = re.compile(r'^(\w|-)+$')
+        for line in self:
+            if not re.match(name_pattern, line.form_field):
+                raise ValidationError(_(
+                    "Invalid form field name %(field_name)s. It should only contain alphanumerics,"
+                    " hyphens or underscores.",
+                    field_name=line.form_field,
+                ))
+            if line.document_type == 'header_footer' and line.form_field.startswith('sol_id_'):
+                raise ValidationError(_(
+                    "Invalid form field name %(field_name)s. A form field name in a header or a"
+                    " footer can not start with \"sol_id_\".",
+                    field_name=line.form_field,
+                ))
+
+    @api.constrains('path')
+    def _check_valid_and_existing_paths(self):
+        """ Verify that the paths exist and are valid.
+
+        :return: None
+        :raises: ValidationError if at least one of the paths isn't valid.
+        """
+        name_pattern = re.compile(r'^(\w|-|\.)+$')
+        for line in self.filtered(lambda line: line.path):
+            if not re.match(name_pattern, line.path):
+                raise ValidationError(_(
+                    "Invalid path %(path)s. It should only contain alphanumerics, hyphens,"
+                    " underscores or points.",
+                    path=line.path,
+                ))
+
+            path = line.path.split('.')
+            is_header_footer = line.document_type == 'header_footer'
+            Model = self.env['sale.order'] if is_header_footer else self.env['sale.order.line']
+            for elem in path:
+                if not Model._fields.get(elem):
+                    raise ValidationError(_(
+                        "The field %(field_name)s doesn't exist on model %(model_name)s",
+                        field_name=elem,
+                        model_name=Model._name
+                    ))
+                if elem != path[-1]:
+                    Model = Model.mapped(elem)
```

Comment:
too magic, use `Model.env[field.comodel_name]` instead

Reply:
??? What's too magic in your opinion here?

Reply:
Maybe too rash, but I feel like `mapped` is to be called on a record(set) (it is one, but represented as model more than a recordset).
Don't you feel it'd be clearer with something like `Model = Model.env[comodel_name]` ?

Reply:
Don't feel so at all... '^^
If it's the same for you, I think `Model = Model.mapped(elem)` is clearer.
Personal preference though, so if there is a rational to change it (perf? I don't feel it'd made any difference, but I might be wrong), I'd prefer to keep it as is

Reply:
Well, I feel like somebody reading it would have to deeply understand the behavior of the module to understand that elem is the name of a relational field, resulting in getting a new Model (aka empty recordset).

Reply:
I could rename elem to something else if you prefer